### PR TITLE
[QUIZ-602] Copy quiz results and send email UI

### PR DESCRIPTION
**Pull request by lws803**
Adds copy quiz results functionality

https://github.com/Constructor-io/constructorio-ui-quizzes/assets/7874219/cb492d59-3954-44c4-ba36-8c743e03a465

### CODE REVIEW

**Comment by lws803** [comment]
@mocca102 I wasn't able to pass `onEmailResults` callback to storybook and e2e tests though unfortunately. Will need some time to look into it but was wondering if you know what the issue is right away 🙂

**Comment by lws803** [comment]
> @mocca102 I wasn't able to pass `onEmailResults` callback to storybook and e2e tests though unfortunately. Will need some time to look into it but was wondering if you know what the issue is right away 🙂

Oh nvm 👍 figured it out here: 4e8c041e63ee0b808f2d79b504b85cadb02b1ecb

File: src/types.ts
Comment (lws803) [reply src/types.ts]: Good call 👍 4e763951c8fb0f85f2587d554ace5ebc47c72f1f

**Review by v-i-s-h-n-u-ps — APPROVED** [review]
@lws803 Thank you. LGTM!

File: package.json
Comment (mocca102) [inline package.json:110]: Sharing my comments about adding these dependencies here as well. We usually try to keep the dependencies few as possible to keep a small size for the bundled version of our library as well. I believe some dependencies might be convenient but some are unnecessary also and don't save us much dev time. I'm tagging @esezen here as well to see what he thinks. 

Sharing gzipped sizes here for easier lookup
"@hookform/resolvers": "923b",
"react-hook-form": "7.4k",
"react-responsive-modal": "3.8k",
"zod": "13.5"

React is 2.7k gzipped for comparison

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (mocca102) [inline src/components/ShareResultsModal/ShareResultsModal.tsx:31]: Can we make this functionality available for the hooks interface as well? Like if people are using the hooks interface they won't have access to this feature. Depending on how quick we want this out I will leave it to you if you agree with me to decide whether to refactor this in this PR to be into a hook (something like useShareResultsLink() or do it in another PR.

File: src/stories/Quiz/argTypes.ts
Comment (mocca102) [inline src/stories/Quiz/argTypes.ts:86]: I believe the `/` at end of line here is extra

File: src/components/ShareResultsModal/EmailField.tsx
Comment (mocca102) [inline src/components/ShareResultsModal/EmailField.tsx:47]: I can imagine this becoming a problem since it indicates that the email has been sent when it actually means that the form has been submitted. Not sure how to handle this though

**Comment by mocca102** [comment]
Are we going to add the handling of the loading of the copied url and rendering the results in a separate PR? @lws803

**Review by mocca102 — COMMENTED** [review]
@lws803 Thanks for working on that awesome work! My only comments are about the used dependencies and refactoring that can wait if we have to get this out sooner.

Will request review from @esezen to see what he thinks

**Comment by lws803** [comment]
> Are we going to add the handling of the loading of the copied url and rendering the results in a separate PR? @lws803

Yes there will be a separate PR for that 😄 https://github.com/Constructor-io/constructorio-ui-quizzes/pull/146

File: src/components/ShareResultsModal/EmailField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:47]: Yeah this is a weird one :x I think what we can do to make this better is to make email submit callback async and only show this message if the callback successfully resolves

Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:47]: ab384b91031ea7727d4bf0bda8d162af1a9b9e6e handleSubmit from react-hook-forms takes an async `onSubmit` function and only changes state to `isSubmitSuccessful` once the promise resolves

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (lws803) [reply src/components/ShareResultsModal/ShareResultsModal.tsx:31]: Oh good call! That won't be hard to move 👍 d280bd3f39a75f56bc6e87fe8e9363857411cf1e let me know if this is what you intended though. Also, do I have to update documentation or ts docs?

File: src/stories/Quiz/argTypes.ts
Comment (lws803) [reply src/stories/Quiz/argTypes.ts]: 030dc0f3119ecfe746210561c897aa72e60855db

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (lws803) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: Oh good call! That won't be hard to move 👍 d280bd3f39a75f56bc6e87fe8e9363857411cf1e let me know if this is what you intended though. Also, do I have to update documentation or ts docs?

File: package.json
Comment (lws803) [reply package.json:110]: I'm still leaning on the side of keeping react-hook-forms but maybe remove zod and hookform resolvers 😛 react-hook-forms helps us with state management which would otherwise be a little more work to track and implement.

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (mocca102) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: Thanks for adding this. We will follow up with a usage example after the new docs PR gets merged

File: src/components/ShareResultsModal/EmailField.tsx
Comment (mocca102) [reply src/components/ShareResultsModal/EmailField.tsx:47]: Nice 🧠

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (mocca102) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: We will follow up with a usage example after the new docs PR gets merged

Comment (mocca102) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: Thanks for refactoring

**Comment by mocca102** [comment]
Thanks for making the changes!

File: package.json
Comment (esezen) [reply package.json:110]: I'm with @mocca102 on this one. With all of our libraries, we take dependencies very seriously and try to have as few of them as possible. Customers also value this highly and even with [client-js](https://github.com/Constructor-io/constructorio-client-javascript) which has only a handful (we are removing 1 dependency soon and 1 in the mid-term), we heard concerns from customers about size. Right now this PR is getting the bundle size above 100KB with a 30% increase in size. 

I believe `@hookform/resolvers` and `zod` are for validation. Considering the amount of validation we do and the complexity, I think we can get rid of these. What do you think? For `react-hook-form`, I'm still in the middle of reviewing the PR so I'm not entirely sure about how it's used. If the work needed to remove it and have our own implementation is around a day's work, I would vote to have our implementation as well. If you think it's more, let's move this to Slack or a to a meeting and discuss internally. There are concerns with dependencies other than size as well.

File: src/components/ShareResultsModal/EmailField.tsx
Comment (esezen) [inline src/components/ShareResultsModal/EmailField.tsx:54]: Not a big deal but it would be better if we follow the existing patterns. We usually put the types to the top of the file as an interface. Can we do that just to keep it consistent? Tagging TS wizard @mocca102 

https://github.com/Constructor-io/constructorio-ui-quizzes/blob/bedaf3e82e855237e2dcef1854a1846c6864a82b/src/components/RedoButton/RedoButton.tsx#L5

File: src/components/ShareResultsModal/LinkField.tsx
Comment (esezen) [inline src/components/ShareResultsModal/LinkField.tsx:33]: Same as above

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (esezen) [inline src/components/ShareResultsModal/ShareResultsModal.tsx:55]: Same as above

File: src/components/ShareResultsModal/EmailField.tsx
Comment (esezen) [inline src/components/ShareResultsModal/EmailField.tsx:44]: If `onEmailResults` callback throws an error, we don't display anything on the page. Is this expected? Do we want the enduser to handle that case? (I'm not against it, just checking)

File: src/components/ShareResultsModal/LinkField.tsx
Comment (esezen) [inline src/components/ShareResultsModal/LinkField.tsx:17]: I believe this has the potential to throw an error. Can we wrap it in try catch?

File: package.json
Comment (lws803) [reply package.json:110]: Oh yeah for sure, I can remove the validator libraries 👍 agreed it makes no sense to blow up our bundle size if we're just using it for one validation (email). I'm on the fence with react-hook-form though since it is pretty helpful in structuring code and maintaining states. It's definitely something I can replace with our own within a day as well but it's nice to have that library manage form states for us 🙂

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (esezen) [inline src/components/ShareResultsModal/ShareResultsModal.tsx:2]: What are the benefits of this library? We already have a modal here: https://github.com/Constructor-io/constructorio-ui-quizzes/blob/main/src/components/SessionPromptModal/SessionPromptModal.tsx

Does this library handle some fancy stuff?

**Review by esezen — COMMENTED** [review]
This is great! Thanks for working on this 🚀 

Left some comments

File: src/components/ShareResultsModal/EmailField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:44]: That's a good question, there are no designs for the error state unfortunately 🤔 but I'll check with Mariano in case he has some thoughts on this.

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (lws803) [reply src/components/ShareResultsModal/ShareResultsModal.tsx:2]: Mostly just modal placement and also handling clicks outside of the modal itself 😄 I can remove this though if you feel strongly about not using this package. Sorry, I should've checked first on the strict requirement to keep bundle sizes small 😅

File: src/components/ShareResultsModal/EmailField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:54]: Whoops, good catch! Sorry, I should've aligned with the existing patterns instead of borrowing practices from another 😛 27f85b661c45a8c98ce347cd87ae97e484c8900e

Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx]: Whoops, good catch! Sorry, I should've aligned with the existing patterns instead of borrowing practices from another 😛 27f85b661c45a8c98ce347cd87ae97e484c8900e

Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx]: Whoops, good catch! Sorry, I should've aligned with the existing patterns instead of borrowing practices from another 😛 75b49eb303d05807142008881e97dc49fef9d339

File: src/components/ShareResultsModal/LinkField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/LinkField.tsx]: 75b49eb303d05807142008881e97dc49fef9d339

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (lws803) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: 75b49eb303d05807142008881e97dc49fef9d339

File: src/hooks/useQuizState/useUrlState.ts
Comment (lws803) [reply src/hooks/useQuizState/useUrlState.ts]: Removed it here 75b49eb303d05807142008881e97dc49fef9d339

File: package.json
Comment (lws803) [reply package.json]: Removed the validators and resolvers here 👍 2d8e9c5387b458dbf10eb8f13ff7e0cff41b364f

File: src/components/ResultContainer/ResultContainer.tsx
Comment (lws803) [reply src/components/ResultContainer/ResultContainer.tsx:40]: That's a good question 🤔 let me check up on that as well

Comment (lws803) [reply src/components/ResultContainer/ResultContainer.tsx]: Added a new prop here 34282f1773f8abcad715fa444f8911fcf63db8f1

File: src/stories/Quiz/argTypes.ts
Comment (lws803) [reply src/stories/Quiz/argTypes.ts:85]: Hmm that's a good suggestion, but I also wanna avoid providing more access unless they specifically ask for it though 🤔 just to keep things on the leaner side

File: src/components/ResultContainer/ResultContainer.tsx
Comment (lws803) [reply src/components/ResultContainer/ResultContainer.tsx:34]: Good call, 3ac51fa9ddcf71f28e5b9ef3891853b83390547b added both classes. `cio-results-filter-and-redo-container` will not technically exist in css but it can be used to preserve backwards compatibility in case folks are using it with selectors

File: src/components/ShareResultsModal/EmailField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:45]: CleanShot 2023-10-31 at 11 19 29@2x

Added an error state here 👍

File: src/components/ShareResultsModal/LinkField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/LinkField.tsx:21]: CleanShot 2023-10-31 at 11 20 08@2x

Added an error for copying as well

Comment (lws803) [reply src/components/ShareResultsModal/LinkField.tsx]: 85a550b81a9df26541cac20a5d7d5af28d154ad3

File: src/components/ShareResultsModal/EmailField.tsx
Comment (lws803) [reply src/components/ShareResultsModal/EmailField.tsx:57]: 85a550b81a9df26541cac20a5d7d5af28d154ad3

File: src/components/ShareResultsModal/ShareResultsModal.tsx
Comment (lws803) [reply src/components/ShareResultsModal/ShareResultsModal.tsx]: Removed modal package 👍 ad9ec88

File: package.json
Comment (esezen) [reply package.json]: For `react-hook-form`, I still think it's better to implement our own instead of bringing a library to handle a single input but let me tag @jjl014 to see what he thinks

Comment (lws803) [reply package.json]: Ahh no worries, if you feel strongly about it, I can remove it as well 👍 it'll just be a few states to maintain on our end

Comment (jjl014) [reply package.json]: Hey Team, I just took a look and I agree with @mocca102 and @esezen here. Generally we try to keep our libraries as lean as possible as we've had complaints from customers before in regards to some of our other libraries. They can be pretty stingy when it comes to size and every byte counts unfortunately. 😓 

From what I can tell looking at the code, it looks like we're using the package for a small portion of the component which can be handled by simple `useState` hooks (but please correct me if I'm wrong). For a package that takes up such a large size of our library, it doesn't seem worth it and probably makes more sense to roll our own.

Comment (lws803) [reply package.json]: Makes sense to me as well 👍 have switched to using our own states. Thanks for clarifying @jjl014 @esezen @mocca102 !

**Review by esezen — APPROVED** [review]
LGTM!

**Pull request merged**
